python/hail/typecheck2/check.py

```diff
@@ -0,0 +1,316 @@
+import sys
+import inspect
+from typing import *
+import collections
+import itertools
+
+implicit_conversions: Dict[type, Set[type]] = {}
+
+warned_for: Set[type] = set()
+
+
+def qualified_name(t):
+    return qualified_type(type(t))
+
+
+def qualified_type(t):
+    module = t.__module__
+    if module == 'builtins':
+        return t.__name__
+    elif module == 'typing':
+        return str(t).replace('typing.', '')  # HACKY HACK
+    else:
+        return t.__qualname__
+
+
+class TypeCheckFailure(Exception):
+    __slots__ = ['msg']
+
+    def __init__(self, msg: str):
+        self.msg = msg
+
+
+def get_spec(f) -> inspect.FullArgSpec:
+    if hasattr(f, '__memo'):
+        return f.__memo
+    else:
+        signature = inspect.getfullargspec(f)
+        f.__memo = signature
+        return signature
+
+
+def typecheck(f: Callable):
+    """Typecheck a function's arguments against its type annotations.
+
+    Examples
+    --------
+    >>> def f(i: int, s: str) -> str:
+    ...     typecheck(f)
+    ...     return i * s
+
+    Notes
+    -----
+    Functions should call this function with themselves as an argument. This
+    is required to look up the function signature without hacking around in
+    the garbage collector and frames too deeply.
+
+    Not all types in :mod:`typing` are supported at this time.
+
+    Parameters
+    ----------
+    f
+        Function to check.
+
+    Raises
+    ------
+    TypeError
+    """
+    frame = inspect.currentframe().f_back
```

Comment:
Docs suggest being careful with frame objects, because they can cause reference cycles. Are we safe here because `frame` is the previous stack frame? Unless you're confident (I don't understand Python well enough) could just be safe and wrap the body of `typecheck` in a
```
try:
    ...
finally:
    del frame
```

Reply:
great suggestion.